libclamav_rust/src/fuzzy_hash.rs

```diff
@@ -0,0 +1,544 @@
+        ))
+    };
+
+    let result: sys::cl_error_t =
+        match fuzzy_hash_load_subsignature(&mut hashmap, hexsig, lsig_id, subsig_id) {
+            Ok(_) => sys::cl_error_t_CL_SUCCESS,
+            Err(error) => {
+                error!("Error when loading fuzzy hash logical subsignature:");
+                error!("  {}", hexsig);
+
+                let result = match error {
+                    FuzzyHashError::Format => {
+                        error!("Invalid signature format.");
+                        sys::cl_error_t_CL_EFORMAT
+                    }
+                    FuzzyHashError::UnknownAlgorithm(algo) => {
+                        error!("Unknown hashing algorithm: {}", algo);
+                        sys::cl_error_t_CL_EFORMAT
+                    }
+                    FuzzyHashError::FormatHammingDistance(dist) => {
+                        error!("Invalid hamming distance: {}", dist);
+                        sys::cl_error_t_CL_EFORMAT
+                    }
+                    FuzzyHashError::FormatHashBytes(hash) => {
+                        error!("Invalid bytes for hash: {}", hash);
+                        sys::cl_error_t_CL_EFORMAT
+                    }
+                    _ => {
+                        error!("Unexpected error");
+                        sys::cl_error_t_CL_ERROR
+                    }
+                };
+
+                error!("Expected format: algorithm#hash[#hammingdistance]");
+                error!("  where");
+                error!("   - algorithm:       Must be 'fuzzy_img'");
+                error!("   - hash:            Must be an 8-byte hex string");
+                error!("   - hammingdistance: (optional) Must be an unsigned integer");
+
+                result
+            }
+        };
+
+    result
+}
+
+/// Load a fuzzy hash subsignature
+/// Parse a fuzzy hash logical sig subsignature.
+/// Add the fuzzy hash to the matcher so it can be matched.
+pub fn fuzzy_hash_load_subsignature(
+    hashmap: &mut Box<HashMap<FuzzyHash, Vec<FuzzyHashMeta>>>,
+    hexsig: &str,
+    lsig_id: u32,
+    subsig_id: u32,
+) -> Result<(), FuzzyHashError> {
+    let mut hexsig_split = hexsig.split('#');
+
+    let algorithm = match hexsig_split.next() {
+        Some(x) => x,
+        None => return Err(FuzzyHashError::Format),
+    };
+
+    let hash = match hexsig_split.next() {
+        Some(x) => x,
+        None => return Err(FuzzyHashError::Format),
+    };
+
+    let distance: u32 = match hexsig_split.next() {
+        Some(x) => match u32::from_str_radix(x, 10) {
+            Ok(n) => n,
+            Err(_) => {
+                return Err(FuzzyHashError::FormatHammingDistance(x.to_string()));
+            }
+        },
+        None => 0,
+    };
+
+    // TODO: Support non-zero distance
+    if distance != 0 {
+        error!(
+            "Non-zero hamming distances for image fuzzy hashes are not supported in this version."
```

Comment:
Perhaps a FLevel thing?

Reply:
I could mention the current FLevel in the log, but there's no way of knowing now which future FLevel will support it.  Telling someone to verify that the FLevel is set correctly implies that some other FLevel will work, and I while I'd like to implement the hamming distance feature in the future, I can't promise if/when that will happen.  I'm inclined to keep this message as-is.